sqlflow_models/gcn.py

```diff
@@ -0,0 +1,369 @@
+            id = inputs['id'].numpy().astype(np.int32)
+            feature = inputs['features'].numpy().astype(np.float32)
+            from_node = inputs['from_node_id'].numpy().astype(np.int32)
+            to_node = inputs['to_node_id'].numpy().astype(np.int32)
+            if int(id) not in ids_check:
+                ids.append(int(id))
+                features.append(feature)
+                labels.append(label.numpy()[0])
+                ids_check[int(id)] = 0
+            if tuple([int(from_node), int(to_node)]) not in edge_check:
+                edge_check[tuple([int(from_node), int(to_node)])] = 0
+                edges.append([from_node, to_node])
+        features = np.stack(features)
+        labels = np.stack(labels)
+        edges = np.stack(edges)
+
+        self.features, self.labels, self.adjacency, self.nshape, na_shape = self.preprocess(ids, features, labels, edges)
+        # training the model
+        wait = 0
+        best_acc = -9999999
+        PATIENCE = 10
+        for epoch in range(self.epochs):
+            # calculate the gradients and take the step
+            loss_value, grads = self.grad(self, [self.features, self.adjacency], self.labels, self.train_mask)
+            optimizer().apply_gradients(zip(grads, self.trainable_variables))
+            # Test on train and evaluate dataset
+            train_acc = self.test(self.train_mask)
+            val_acc = self.test(self.val_mask)
+            print("Epoch {} loss={:6f} accuracy={:6f} val_acc={:6f}".format(epoch, loss_value, train_acc, val_acc))
+            # early stopping
+            if epoch > 50 and self.early_stopping:
+                if float(val_acc.numpy()) > best_acc:
+                    best_acc = float(val_acc.numpy())
+                    wait = 0
+                else:
+                    if wait >= PATIENCE:
+                        print('Epoch {}: early stopping'.format(epoch))
+                        break
+                    wait += 1
+        # evaluate the model
+        result = self.evaluate(data=[self.features, self.adjacency], y=self.labels, sample_weight=self.val_mask)
+        # get all the results
+        predicted = self.predict([self.features, self.adjacency])
+        # store the results in a pickled file
+        with open('./results.p', 'wb') as f:
+            results = dict()
+            for i in range(len(ids)):
+                results[ids[i]] = predicted[i]
+            results['evaluation'] = result
+            pickle.dump(results, f)
+
+    def sqlflow_evaluate_loop(self, x, metric_names):
+        """Customed evaluation, can only support calculating the accuracy."""
+        with open('./results.p', 'rb') as f:
+            results = pickle.load(f)
+            eval_result = results['evaluation']
+        return eval_result
+
+    def sqlflow_predict_one(self, sample):
+        """Customed prediction, sample must be the node id."""
+        with open('./results.p', 'rb') as f:
```

Comment:
we don't need to open the file each time the sqlflow_predict_one is called

Reply:
Yeah sure! This is indeed inefficient, but how should we retrieve the result from the file during prediction?

Reply:
Can we try loading the file in `__init__` if the file exists?

Reply:
Yes, this is what I'm thinking rn. If sqlflow will init the model once during the evaluation and prediction phase, then we could load the file in the `__init__` and use it directly.